Frameworks/Foundation/NSUndoManager.mm

```diff
@@ -38,67 +102,159 @@ - (id)prepareWithInvocationTarget:(id)target {
+        [self endUndoGrouping];
+    }
+
+    [self undoNestedGroup];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)undoNestedGroup {
-    UNIMPLEMENTED();
+    _groupingLevel = [_undoStack count];
+
+    if (![self canUndo]) {
+        return;
+    }
+
+    if (_groupingLevel == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"No top level group to undo."];
+    }
+
+    _undoing = YES;
+
+    _NSUndoGroup* undoGroup = [_undoStack peek];
+
+    // Record redo operation as a single redo group
+    [self _beginRedoGrouping];
+    [undoGroup undo];
+    [self _endRedoGrouping];
+    [_undoStack pop];
+
+    _groupingLevel = [_undoStack count];
+    ;
+
+    _undoing = NO;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)redo {
-    UNIMPLEMENTED();
+    _groupingLevel = [_redoStack count];
+    if (![self canRedo]) {
+        return;
+    }
+
+    if (_undoing) {
+        [NSException raise:NSInternalInconsistencyException format:@"Cannot invoke redo while undoing is active."];
+    }
+
+    _redoing = YES;
+
+    _NSUndoGroup* redoGroup = [_redoStack peek];
+
+    // Record undo operation as a single group
+    [self beginUndoGrouping];
+    [redoGroup undo];
+    [self endUndoGrouping];
+    [_redoStack pop];
+
+    _groupingLevel = [_redoStack count];
+
+    _redoing = NO;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)beginUndoGrouping {
-    UNIMPLEMENTED();
+    _NSUndoGroup* undoGroup = [_undoStack peek];
+
+    if (undoGroup == nil || [undoGroup isClosed]) {
+        _NSUndoGroup* newGroup = [[_NSUndoGroup alloc] initWithLevel:1];
+        [_undoStack push:newGroup];
+    } else {
+        [undoGroup createUndoGroupWithLevel:1];
+    }
+    _groupingLevel = [_undoStack count];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)endUndoGrouping {
-    UNIMPLEMENTED();
+    // No group to close.
+    if ([_undoStack count] == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"No group to close."];
+    }
+    _NSUndoGroup* undoGroup = [_undoStack peek];
+    [undoGroup closeUndoGroup];
+}
+
+// Internal redo grouping
+- (void)_beginRedoGrouping {
+    _NSUndoGroup* redoGroup = [_redoStack peek];
+
+    if (redoGroup == nil || [redoGroup isClosed]) {
+        _NSUndoGroup* newGroup = [[_NSUndoGroup alloc] initWithLevel:1];
+        [_redoStack push:newGroup];
+    } else {
+        [redoGroup createUndoGroupWithLevel:1];
+    }
+    _groupingLevel = [_redoStack count];
+}
+
+- (void)_endRedoGrouping {
+    // No group to close.
+    if ([_redoStack count] == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"No redo group to close."];
+    }
+    _NSUndoGroup* redoGroup = [_redoStack peek];
+    [redoGroup closeUndoGroup];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)disableUndoRegistration {
-    UNIMPLEMENTED();
+    _disableCount++;
```

Comment:
That's actually the point believe it or not. Multiple disable calls are recorded. Enable calls must equal the amount of disable calls when invoking any undo operation. Otherwise, an inconsistency exception is raised.

---
In reply to: [76160566](https://github.com/Microsoft/WinObjC/pull/900#discussion_r76160566) [](ancestors = 76160566)